src/transformers/models/audioflamingo3/modular_audioflamingo3.py

```diff
@@ -0,0 +1,517 @@
+            )
+            return tuple(v for v in outs if v is not None)
+
+        return BaseModelOutput(
+            last_hidden_state=h,
+            hidden_states=tuple(hs_list) if hs_list is not None else None,
+            attentions=tuple(attn_list) if attn_list is not None else None,
+        )
+
+
+class AudioFlamingo3MultiModalProjector(nn.Module):
+    """
+    Audio adaptor (a small MLP) that projects AudioFlamingo3Encoder (AF-Whisper)
+    features to the LLM embedding space so they can replace `<sound>` tokens.
+    """
+
+    def __init__(self, config: AudioFlamingo3Config) -> None:
+        super().__init__()
+        d_audio = config.audio_config.d_model
+        d_text = config.text_config.hidden_size
+        self.layers = nn.ModuleList([nn.Linear(d_audio, d_text), nn.GELU(), nn.Linear(d_text, d_text)])
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        for layer in self.layers:
+            x = layer(x)
+        return x
+
+
+@auto_docstring(
+    custom_intro="""
+    The AudioFlamingo3 model which consists of a audio backbone and a language model.
+    """
+)
+class AudioFlamingo3ForConditionalGeneration(AudioFlamingo3PreTrainedModel, GenerationMixin):
+    """
+    AudioFlamingo3 model composed of an audio encoder, a projection to the LM hidden size, and a causal LM.
+
+    The audio-text fusion is performed by *replacing* occurrences of the `<sound>` token with per-frame audio embeddings,
+    without changing sequence length. The number of `<sound>` tokens is expected to match the *post-pool* frame count
+    computed by the processor.
+    """
+
+    config_class = AudioFlamingo3Config
+
+    def __init__(self, config: AudioFlamingo3Config):
+        super().__init__(config)
+        # Language model
+        self.language_model = AutoModelForCausalLM.from_config(config.text_config)
+        # Audio encoder (explicitly instantiate our class to guarantee helper availability)
+        self.audio_tower = AudioFlamingo3Encoder(config.audio_config)
+        # Projection to LM hidden size
+        self.multi_modal_projector = AudioFlamingo3MultiModalProjector(config)
+
+        self.post_init()
+
+    # --- Embedding plumbing (forward to LM) ---
+    def get_input_embeddings(self):
+        return self.language_model.get_input_embeddings()
+
+    def set_input_embeddings(self, value):
+        self.language_model.set_input_embeddings(value)
+
+    def get_output_embeddings(self):
+        return self.language_model.get_output_embeddings()
+
+    def set_output_embeddings(self, value):
+        self.language_model.set_output_embeddings(value)
+
+    def set_decoder(self, decoder):
+        self.language_model.set_decoder(decoder)
+
+    def get_decoder(self):
+        return self.language_model.get_decoder()
+
+    @auto_docstring
+    def forward(
+        self,
+        input_ids: Optional[torch.LongTensor] = None,
+        input_features: Optional[torch.FloatTensor] = None,  # (#windows, n_mels, T_mel)
```

Comment:
_note to self: change to `audio_spectrogram` for more intuitive input as per standardization for audio models?_